rust/feature-flags/src/cohorts/cohort_operations.rs

```diff
@@ -0,0 +1,684 @@
+use common_types::PersonId;
+use petgraph::algo::is_cyclic_directed;
+use petgraph::algo::toposort;
+use petgraph::graph::DiGraph;
+use serde_json::Value;
+use sqlx::Row;
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::collections::VecDeque;
+use std::sync::Arc;
+use tracing::instrument;
+
+use crate::cohorts::cohort_models::{Cohort, CohortId, CohortProperty, InnerCohortProperty};
+use crate::flags::flag_matching::PostgresReader;
+use crate::properties::property_matching::match_property;
+use crate::properties::property_models::OperatorType;
+use crate::{
+    api::errors::FlagError, client::database::Client as DatabaseClient,
+    properties::property_models::PropertyFilter,
+};
+
+use super::cohort_models::CohortPropertyType;
+use super::cohort_models::CohortValues;
+
+impl Cohort {
+    /// Returns all cohorts for a given team
+    #[instrument(skip_all)]
+    pub async fn list_from_pg(
+        client: Arc<dyn DatabaseClient + Send + Sync>,
+        project_id: i64,
+    ) -> Result<Vec<Cohort>, FlagError> {
+        let mut conn = client.get_connection().await.map_err(|e| {
+            tracing::error!("Failed to get database connection: {}", e);
+            FlagError::DatabaseUnavailable
+        })?;
+
+        let query = r#"
+            SELECT c.id,
+                  c.name,
+                  c.description,
+                  c.team_id,
+                  c.deleted,
+                  c.filters,
+                  c.query,
+                  c.version,
+                  c.pending_version,
+                  c.count,
+                  c.is_calculating,
+                  c.is_static,
+                  c.errors_calculating,
+                  c.groups,
+                  c.created_by_id
+              FROM posthog_cohort AS c
+              JOIN posthog_team AS t ON (c.team_id = t.id)
+            WHERE t.project_id = $1
+            AND c.deleted = false
+        "#;
+        let cohorts = sqlx::query_as::<_, Cohort>(query)
+            .bind(project_id)
+            .fetch_all(&mut *conn)
+            .await
+            .map_err(|e| {
+                tracing::error!("Failed to fetch cohorts from database: {}", e);
+                FlagError::Internal(format!("Database query error: {}", e))
+            })?;
+
+        Ok(cohorts)
+    }
+
+    /// Parses the filters JSON into a CohortProperty structure
+    // TODO: this doesn't handle the deprecated "groups" field, see
+    // https://github.com/PostHog/posthog/blob/feat/dynamic-cohorts-rust/posthog/models/cohort/cohort.py#L114-L169
+    // I'll handle that in a separate PR.
```

Comment:
That link 404s. Is this comment still relevant?

Reply:
sorta; link 404s because I attached it to a branch diff instead of master 🤦 

I'll keep the comment in because it's still relevant although not required – it's more completionist.